### ISSUE TITLE
Migrate to Jakarta EE

### ISSUE THREAD

**Issue by albfernandez** [albfernandez/richfaces#81]
This issue aggregate al tasks necessaries to complete the migration to Jakarta EE

**Tasks**

**jsf-test**

- [ ] albfernandez/jsf-test#14
- [ ] albfernandez/jsf-test#15

**richfaces-cdk**

- [ ] albfernandez/richfaces-cdk#15
 
**richfaces**

- [ ] albfernandez/richfaces-cdk#80
- [ ] albfernandez/richfaces-cdk#81

**Comment by melloware** [comment]
@albfernandez you may also want to rip out the CKEditor component while I think the license is perfectly fine the PrimeFaces Extensions project got hounded by their lawyers until we just ripped it out.  See: https://github.com/primefaces-extensions/primefaces-extensions/issues/1106

**Comment by miguelzapaton** [comment]
Anybody has started with the Jakarta EE migration yet?

**Comment by miguelzapaton** [comment]
I just did a work-in-progress commit to share an intermediate result (with MZ comments, println logging, etc.) of my Jakarta EE migration intent:

* versioned as 10.0.0-SNAPSHOT to align with ee10
* currently myfaces only
* builds a jar (when tests are bypassed)

Most important issues:
* tests with errors or failing, especially: core, a4j, rich
* managed beans are removed, CDI environment required

Artifacts:

https://github.com/miguelzapaton/jsf-test

https://github.com/miguelzapaton/richfaces-cdk

https://github.com/miguelzapaton/richfaces


Hints to try it out:

1) mvn clean install jsf-test (with tests, resolve workspace artifacts, update snapshots)

2) mvn clean install richfaces-cdk (with tests, resolve workspace artifacts, update snapshots)

3) mvn clean install only the core project of richfaces (with tests, resolve workspace artifacts, update snapshots)

4) mvn install richfaces (currently skip tests, resolve workspace artifacts, update snapshots)

**Feedback (and help) very welcome!**

**Comment by melloware** [comment]
Nice work @miguelzapaton !   Making great progress!

**Comment by albfernandez** [comment]
Great work @miguelzapaton 

I like the 10.0 version idea.

Maybe setting java version to 11 (and ensuring it works with 11,17,21)

I think next step should be migrate examples/showcase project to manual testing

Note to me (and others)

```bash
jsf-test:           mvn clean install
richfaces-cdk:      mvn clean install
richfaces (core):   mvn clean -DskipTests install
richfaces(root):    mvn clean -Dmaven.test.skip install
```

**Comment by albfernandez** [comment]
I've created the jakarta branch for jsf-test, richfaces-cdk and richfaces in my repository and incorporated your changes

**Comment by MilovdZee** [comment]
I have been working on it and fixed some issues. In my project that uses richfaces most is now working: https://github.com/MilovdZee/richfaces-jakarta/tree/upgrading_to_jakarta

Problem now is that http://localhost:8080/org.richfaces.resources/jakarta.faces.resource/org.richfaces/richfaces.js is not found. I did enable the resource mapper and resource servlet but somehow it does not work. When I copy all the richfaces resources to my own project in the resources folder it does work. Could it be that JSF changed in that it does not allow downloading resources from dependencies?

**Comment by melloware** [comment]
Jakarta JSF still allows resources the same as JavaX version no change there.

**Comment by MilovdZee** [comment]
Oke, then there must be something else why it can't find the resources. I traced it a lot and at some point it just forwards the resource request to the default one after chopping of the `org.richfaces.resources` part. And then it is not found.

**Comment by MilovdZee** [comment]
I'm very stuck. In my application richfaces tries to request `http://localhost:8080/org.richfaces.resources/jakarta.faces.resource/org.richfaces/richfaces-base-component.js` but it does not find the resource. I see it entering `org.richfaces.webapp.ResourceServlet.httpService()` and it forwards the request to `facesServlet.service(request, response)`but then nothing is found. 
Any suggestion is very welcome. This looks like one of the last hurdles to get this really useful.

**Comment by melloware** [comment]
your URL looks weird?  Here is what PrimeFaces looks like asking for core.js

```
https://www.primefaces.org/showcase/jakarta.faces.resource/core.js.xhtml?ln=primefaces
```

so yours should not have a JS extension?   Why is `org.richfaces` after resource but before the JS?  the `ln=primefaces` i would expect to be `ln=richfaces` for standard JSF?

**Comment by MilovdZee** [comment]
Agreed. But this is how it used to be. In a working application that uses the javax original richfaces the request looks like this: `https://<host>/org.richfaces.resources/javax.faces.resource/org.richfaces/richfaces.js`

**Comment by melloware** [comment]
ok good.  Weird...but good.

**Comment by MilovdZee** [comment]
From what I understand the `org.richfaces.resources` is to map it to the resourcehandler instead of by using .xhtml or .jsf. And the `org.richfaces`is handled as an alternative to the ln= part. Should be fine and the request is actually picked up by the resourceHandler.

**Comment by melloware** [comment]
so then in the JAR file i would expect `META-INF/resources/org.richfaces/richfaces.js` to be there can you confirm that file is there exactly in your Richfaces JAR?

**Comment by MilovdZee** [comment]
Yes, I checked the jar and the files are there

**Comment by MilovdZee** [comment]
![image](https://github.com/user-attachments/assets/eefc65c1-e053-46c2-9892-23beff20f0a6)

**Comment by melloware** [comment]
OK then i am stumped...

**Comment by MilovdZee** [comment]
https://issues.apache.org/jira/browse/MYFACES-4524
This creates the problem. In FacesServletMappingUtils.createMappingFromServletRegistration() is the servletPath matched exactly then it used to return a prefixMapping but now it returns an exactMapping and this is later fully ignored. Probably I can update the servlet org.richfaces.webapp.ResourceServlet to trigger a prefixMapping

**Comment by MilovdZee** [comment]
Updating org.richfaces.webapp.ResourceServlet fixed the problem of not fining resources:
It was `public class ResourceServlet implements Servlet {}` and by adding `DelegatedFacesServlet`ir now creates the correct mapping.
`public class ResourceServlet implements Servlet, DelegatedFacesServlet {}`

**Comment by MilovdZee** [comment]
Next issue... jsf.js does not seem to exist anymore and RichFaces uses it in it's js files

**Comment by melloware** [comment]
yep `faces.js`.  And yeah MyFaces issue probably closed a gap in the spec.

**Comment by MilovdZee** [comment]
I updated the javascript to use faces instead of jsf and all seems to work fine now :)

Maybe time to merge my changes into the main jakarta branch. I'll try to create a fork and then a merge request

**Comment by MilovdZee** [comment]
skin does not work

**Comment by miguelzapaton** [comment]
Really cool what my eyes are seeing here @MilovdZee

Thank you @albfernandez for creating the jakarta branch and using my intermediate result as a starting point. 
Currently I'm too busy with other things but I hope to have time for further improvements soon. I still have some entries on my list regarding the tests etc.

Following the discussion above I'm a bit skeptical about the editor and mediaOutput components. I think there are strong reasons to no longer include them and maybe that's the case for other components too.

@melloware Do you think that a well running Jakarta RF 10 could be run in parallel with Primefaces without further adaptions? 
I (and maybe others) would like to do an inplace migration in one of my applications.

Looking forward to getting this fully done together.

**Comment by melloware** [comment]
As for PF it should play just fine with this Jakarta RF branch!  

I really think the CKeditor component should be removed as I mentioned they harassed PF extensions project about licensing until we finally removed it.

**Comment by MilovdZee** [comment]
I don't care about the CKeditor (I'm using TinyMCE) but I do need the mediaOutput. What is the reason not to keep the mediaOutput? If needed we can introduce a flag that enables the component.

**Comment by albfernandez** [comment]
mediaoutput was removed to fix CVE-2018-12532, If reintroduced, this vulnerability have to be fixed first.

ckeditor has two issues:
- The license issue.  I think Richfaces license is compatible, but if they are harassing other opensource projects it's safe to remove. 
- Ckeditor 4 (open source) is not maintained anymore.

**Comment by MilovdZee** [comment]
`
@ManagedBean("richfacesVersion")
@ApplicationScoped
public class RichfacesVersion {
    public String getVersion() {
        return "richfaces version";
    }
}
`

This does not work. If I replace @ManagedBean with @Named then it works. This is probably the reason that the RF beans are not picked up. They are defined as managedbean in the core.faces-config.xml. I checked that the core.faces-config.xml file is scanned and it is. When I create a classname fault I do indeed get an error. So the 'managed-bean' entries are not read somehow.

I also tried annotating the beans in RF with @Named but even then they are not picked up. Any suggestions?

**Comment by melloware** [comment]
Yes ManagedBeans are gone you have to use CDI and @Named

**Comment by MilovdZee** [comment]
That is indeed what I remember but why is @ Named not  being picked up when I use it in the RF bean definitions?

**Comment by MilovdZee** [comment]
That also solved the skinning problem

**Comment by MilovdZee** [comment]
The response looks fine:
`
<?xml version="1.0" encoding="UTF-8"?>
  <partial-response id="j_id__v_0">
    <changes>
      <update id="messageTabBody:newMessage:newMessagePanel">
        <![CDATA[<div id="messageTabBody:newMessage:newMessagePanel"....
`

**Comment by melloware** [comment]
what version of MyFaces and Mojarra are you testing with?  MyFaces 4.0.2 and Mojarra 4.0.9?

**Comment by MilovdZee** [comment]
Yes 4.0.2 and no Mojarra

I also see that the response is exactly the same in a working environment:
```
<?xml version="1.0" encoding="UTF-8"?>
  <partial-response id="j_id__v_0">
    <changes>
      <update id="messageTabBody:newMessage:newMessagePanel">
        <![CDATA[<div id="messageTabBody:newMessage:newMessagePanel" 
```

**Comment by MilovdZee** [comment]
After comparing I see that after the render the failing one is actually updated but with visibility set to hidden and the height and width are not updated. So this is probably an issue with the RF popupPanel.js. It probably needs to update some values after rendering.

**Comment by melloware** [comment]
@MilovdZee if its not calling `show()` is sounds like the AJAX oncomplete or something is not working.  Any error in F12 console?

**Comment by MilovdZee** [comment]
No errors at all. I'm now calling the show() after a window.setTimeout() of 10ms and it works fine :)
Probably the dom was not ready yet and in the old version was a bit quicker or something. 
This might not have to do anything with RF. This might me abusing RF in a way it should not be used.

**Comment by MilovdZee** [comment]
![image](https://github.com/user-attachments/assets/0d8a1357-6e86-416f-9171-79bc9353f0fb)

With a lot of the tests. I disable all tests for the 'rich components' module and some scattered around but many are already working fine

**Comment by MilovdZee** [comment]
Might it be an idea to just integrate CDK and jsf-test into this project instead of having them external?

**Comment by albfernandez** [comment]
> Might it be an idea to just integrate CDK and jsf-test into this project instead of having them external?

I think there should be separate projects, I've jboss-seam fork that relays on richfaces-cdk

**Comment by MilovdZee** [comment]
Really? Oke, yes, if it is actually used elsewhere then it need to be separate.

For the tests to work we do need some serious fixing though.

**Comment by MilovdZee** [comment]
My application now does a lot more full page reloads. So that is annoying. Something with the rendering changed...

**Comment by MilovdZee** [comment]
I tested a lot and I don't see any issues any more. 
Except for that I now see a bit more full-page renders that result in a flickering screen. But that is only at moments I actually switch pages and so might be expected. But it used to be nicer. And I think this issue comes from MyFaces and not RichFaces. I investigated the response xml and they are 100% the same between the old Richfaces with the old Myfaces and this newly migrated version with a much newer Myfaces. So it must be in the handling of the received xml page fragment. And that is handled by Myfaces.

But let's merge :)
https://github.com/albfernandez/richfaces/pull/85

**Comment by albfernandez** [comment]
Good work! Thanks.
I'll try to review and test the changes this weekend

I'm worried about mediaoutput... CVE-2018-12532 has to be fixed before release.

**Comment by MilovdZee** [comment]
Agreed. Maybe just do it the easy way and make it opt-in functionality with a big warning? That gives us (you?) some time to fix it.

And I disabled quite some tests...

**Comment by MilovdZee** [comment]
The flickering was due to that I used 'redirect' in the faces-config. I now removed that and that now is fine.

**Comment by MilovdZee** [comment]
But there is a bug that looks very related to MyFaces but I don't know:
in `status.js`:
`    const statusAjaxEventHandler = function (data, methodName) {
        if (methodName) {
            //global status name
            const statusName = getGlobalStatusNameVariable();

            let statusApplied = false;
            const statusDataAttribute = getStatusDataAttributeName(statusName);

            let statusContainers;
            if (statusName) {
                statusContainers = [$(document)];
            } else {
                const source = data.source;
                if (source === null) {
                    console.log(`Error: source is null. methodName='${methodName}', statusName='${statusName}'`)
                    return;
                }

                // the element reference will be stale if it was rerendered
                // we need to find the current element in the DOM
                const currentSource = document.getElementById(source.id);

                statusContainers = [$(currentSource).parents('form'), $(document)];
`

**Comment by MilovdZee** [comment]
This piece of code often shows the warning. Not sure why. Does not give any problems though
```
    public void addQueue(String clientName, UIComponent component) {
        if (!containsQueue(clientName)) {
            queuesData.put(clientName, component);
        } else {
            LOGGER.warn("Queue with name '" + clientName + "' has already been registered");
        }
    }
```

**Comment by MilovdZee** [comment]
> Good work! Thanks. I'll try to review and test the changes this weekend
> 
> I'm worried about mediaoutput... [CVE-2018-12532](https://github.com/advisories/GHSA-3hx6-fqpj-xfjr) has to be fixed before release.

Any luck yet? I deployed the application to production and a team now has been using it for about a week. I did not get any feedback about things that are not working. At least not related to the RichFaces upgrade.

**Comment by miguelzapaton** [comment]
@MilovdZee Did you get the showcase (examples) running?

**Comment by MilovdZee** [comment]
I did not look at the showcase

**Comment by MilovdZee** [comment]
There is an issue in myfaces that prevents a4j:status to function properly. See https://issues.apache.org/jira/browse/MYFACES-4712

**Comment by MilovdZee** [comment]
Starting work on MediaOutput: "RichFaces vulnerable to Expression Language Injection". See https://github.com/advisories/GHSA-3hx6-fqpj-xfjr

**Comment by MilovdZee** [comment]
Any suggestions on how to fix this?

Maybe we can sign the request? In normal situations RichFaces created the query string and so is in full control. That string is fine and can be used freely. If we can be sure that the string is created by RF all is fine? This can be accomplished by signing the string. At boot we could create a key and sign all constructed 'do' query strings. A hacker won't know the key and so is unable to inject a accepted query.

**Comment by melloware** [comment]
@MilovdZee yes here is how PrimeFaces fixed it years ago.  You have to use the SessionMap

https://github.com/primefaces/primefaces/commit/26e44eb7962cbdb6aa2f47eca0f230f3274358f0

**Comment by albfernandez** [comment]
@melloware  SessionMap is the safest option

**Comment by albfernandez** [comment]
Hi @MilovdZee , take a look at 23371427c1e57e945ec25cf3d07d44c6d99e009b

This change will encrypt al resource data (mediaoutput and others)

It needs some more testing but can be a good starting point.

The solution proposed by @melloware  is safer, maybe first with encrypting all request and for the next version investigate sessionMap

**Comment by melloware** [comment]
It's what fixed PrimeFaces years ago for the same problem and we have passed multiple penetration tests in the years since.  It takes the value off of the URL so it can't be tampered with

**Comment by MilovdZee** [comment]
I do indeed prefer the PrimeFaces solution where the data is not even in the call. Just a reference. The itching thing is that I'm not sure if nobody finds a way to manipulate the SessionMap. You do also encrypt the key so PrimeFaces has two layers of security. Very nice.
I think for RichFaces the encryption of the data itself is also fine and much closer what is implemented already. So yes, @albfernandez , I do like your solution. looks solid.

**Comment by MilovdZee** [comment]
The setPassword in Codec does not do anything and so even though you try to encrypt it now it still is plain:
```
              Cipher encripter = Cipher.getInstance("AES");
              encripter.init(Cipher.ENCRYPT_MODE, secretKey);
              
              Cipher decripter = Cipher.getInstance("AES");
              decripter.init(Cipher.DECRYPT_MODE, secretKey);
```
It uses local variables and so the global encripter and decripter still will be null.

**Comment by MilovdZee** [comment]
And maybe just set the password in the Codec constructor. There is no need for unencrypted coding anyway

**Comment by MilovdZee** [comment]
I replaced the Codec with this and all still seems to work. I'm not very happy with the algorithm but maybe strong enough?

```
/**
 * License Agreement.
 * <p>
 * Rich Faces - Natural Ajax for Java Server Faces (JSF)
 * <p>
 * Copyright (C) 2007 Exadel, Inc.
 * <p>
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 * <p>
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * <p>
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
 */
package org.ajax4jsf.util.base64;

import jakarta.faces.FacesException;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;

public class Codec {
    private static final String ALGORITHM = "AES";
    public static final int AES_KEY_SIZE = 128; // in bits

    private Cipher decripter;
    private Cipher encripter;

    public Codec() {
        super();

        initCiphers();
    }

    private void initCiphers() throws FacesException {
        try {
            SecureRandom random = SecureRandom.getInstanceStrong();
            KeyGenerator keyGen = KeyGenerator.getInstance("AES");
            keyGen.init(AES_KEY_SIZE, random);
            SecretKey secretKey = keyGen.generateKey();

            encripter = Cipher.getInstance(ALGORITHM);
            encripter.init(Cipher.ENCRYPT_MODE, secretKey);

            decripter = Cipher.getInstance(ALGORITHM);
            decripter.init(Cipher.DECRYPT_MODE, secretKey);
        } catch (Exception e) {
            throw new FacesException("Error set encryption key", e);
        }
    }

    public String decode(String str) throws Exception {
        byte[] src = str.getBytes(StandardCharsets.UTF_8);
        byte[] utf8 = decode(src);
        return new String(utf8, StandardCharsets.UTF_8);
    }

    public String encode(String str) throws Exception {
        byte[] src = str.getBytes(StandardCharsets.UTF_8);
        byte[] utf8 = encode(src);
        return new String(utf8, StandardCharsets.UTF_8);
    }

    public byte[] decode(byte[] src) throws Exception {
        byte[] dec = URL64Codec.decodeBase64(src);
        return decripter.doFinal(dec);
    }

    public byte[] encode(byte[] src) throws Exception {
        byte[] dec = encripter.doFinal(src);
        return URL64Codec.encodeBase64(dec);
    }
}
```

**Comment by MilovdZee** [comment]
Nice. The 'do' for the same resource is now different every time:
```
https://xxx/rfRes/org.richfaces.resource.MediaOutputResource.xhtml?do=728!Mt8lyiSbvUNXfoHL9fWiPTjNgqO92J...
https://xxx/rfRes/org.richfaces.resource.MediaOutputResource.xhtml?do=v5vjyGjd8xvYv0yop4tBYai7pCOPVKD6P...
https://xxx/rfRes/org.richfaces.resource.MediaOutputResource.xhtml?do=rS8Xk1ZkODqUJQ8F0WCeitr!JocQTFO...
```

**Comment by MilovdZee** [comment]
Would be nice if every session gets it's own Codec instance. Currently it is application scoped due to it being a static field

**Comment by albfernandez** [comment]
@MilovdZee I've committed changes , It's similar at your idea, but I want encryption to be configurable with two init parameters:
org.richfaces.ENCRYPT_RESOURCE_DATA  default true, enabled or disabled encryption
org.richfaces.ENCRYPT_PASSWORD to set the password (default random generated). If set can be used in distributed server environments, and also to survive reboots. (it's more secure to leave default value)

In [Richfaces 3.3. documentation](https://docs.jboss.org/richfaces/latest_3_3_X/en/devguide/html/SettingsForDifferentEnvironments.html)  this parameters exist, but seems they never really implemented them:
org.ajax4jsf.ENCRYPT_RESOURCE_DATA
org.ajax4jsf.ENCRYPT_PASSWORD


Next steep (someday): store in sessionMap as proposed by @melloware  (I think it will be the same effort as one codec for session, and it's more secure)

I think with these changes we fix [CVE-2018-12532](https://nvd.nist.gov/vuln/detail/CVE-2018-12532) without removing mediaoutput

**Comment by MilovdZee** [comment]
I saw your changes but the encryption had to be fixed. Did you adopt any of my changes?

**Comment by albfernandez** [comment]
some fixed are incorporated, for example the private fields encripter and decripter are properly initialized now.
Changes in encryption algorithm are not included.
If changes in encryption are for a more secure algorithm, we must check how to use this algorithms with text passwords.

**Comment by MilovdZee** [comment]
Great. I'll update the Codec to support passwords and enablement like you indicated.

**Comment by albfernandez** [comment]
> Great. I'll update the Codec to support passwords and enablement like you indicated.

I've done in this fork. Only changes not adopted the algortihm and initialization vector for password

**Comment by MilovdZee** [comment]
Exactly. I'm now changing to the secure encryption and the two parameters. Currently testing

**Comment by MilovdZee** [comment]
I noticed that you are mainly working on the showcase so won't bite :)

**Comment by MilovdZee** [comment]
Done. I created a new PR with the CODEC changes: https://github.com/albfernandez/richfaces/pull/90

It now uses:
```xml
    <context-param>
        <!-- default is 'true' -->
        <param-name>org.ajax4jsf.ENCRYPT_RESOURCE_DATA</param-name>
        <param-value>true</param-value>
    </context-param>
<!--    <context-param>-->
            <!-- default is random password -->
<!--        <param-name>org.ajax4jsf.ENCRYPT_PASSWORD</param-name>-->
<!--        <param-value></param-value>-->
<!--    </context-param>-->
```

**Comment by MilovdZee** [comment]
Thanks for merging. And I see some 'release' commits. Are you planning to merge into main?

**Comment by albfernandez** [comment]
> Thanks for merging. And I see some 'release' commits. Are you planning to merge into main?

I've released jsf-test 10 and richfaces-cdk 10 to maven. 
The changes in richfaces are to point to the stable releases.

The plan is maintain two branches
main for javax
jakarta for jakarta
